crates/holochain/src/test_utils/consistency.rs

```diff
@@ -0,0 +1,479 @@
+            if cell_id.dna_hash() != space {
+                continue;
+            }
+            let space = spaces
+                .entry(cell_id.dna_hash().clone())
+                .or_insert_with(|| vec![None; conductors.len()]);
+            if space[i].is_none() {
+                let p2p_env: EnvRead = c.get_p2p_env(cell_id.dna_hash().to_kitsune()).await.into();
+                space[i] = Some((p2p_env, Vec::new()));
+            }
+            space[i].as_mut().unwrap().1.push((
+                // TODO: Handle error.
+                c.get_cell_env_readonly(&cell_id).await.unwrap(),
+                cell_id.agent_pubkey().to_kitsune(),
+            ));
+        }
+    }
+    let all_hashes = hashes
+        .into_iter()
+        .map(|h| h.to_kitsune())
+        .collect::<Vec<_>>();
+    for (_, conductors) in spaces {
+        let mut wait_for_agents = HashSet::new();
+        let (tx, rx) = tokio::sync::mpsc::channel(1000);
+        for c in conductors {
+            if let Some((p2p_env, agents)) = c {
+                wait_for_agents.extend(agents.iter().map(|(_, agent)| agent.clone()));
+                tokio::spawn({
+                    let all_hashes = all_hashes.clone();
+                    let tx = tx.clone();
+                    let agent_env_map: HashMap<_, _> =
+                        agents.iter().cloned().map(|(e, a)| (a, e)).collect();
+                    async move {
+                        let mut all_agents = Vec::with_capacity(agents.len());
+                        for (_, agent) in &agents {
+                            // TODO: Handle error.
+                            if let Some(storage_arc) = request_arc(&p2p_env, &agent).await.unwrap()
+                            {
+                                all_agents.push((agent.clone(), storage_arc));
+                            }
+                        }
+                        let iter =
+                            generate_session(&all_agents, &all_hashes, timeout, agent_env_map);
+                        check_all(iter, tx, p2p_env).await;
+                    }
+                });
+            }
+        }
+        wait_for_consistency(rx, wait_for_agents, timeout.clone()).await;
+    }
+}
+
+impl Reporter {
+    async fn send_report(&self, report: SessionReport) {
+        if self
+            .0
+            .send(SessionMessage {
+                from: self.1.clone(),
+                report,
+            })
+            .await
+            .is_err()
+        {
+            tracing::error!("Failed to message for consistency session");
+        }
+    }
+}
+
+async fn wait_for_consistency(
+    mut rx: tokio::sync::mpsc::Receiver<SessionMessage>,
+    mut agents: HashSet<Arc<KitsuneAgent>>,
+    timeout: Duration,
+) {
+    let deadline = tokio::time::Instant::now() + timeout + Duration::from_millis(100);
+    while let Ok(Some(SessionMessage { from, report })) =
+        tokio::time::timeout_at(deadline, rx.recv()).await
```

Comment:
Also not sure about this use of `deadline`, I would expect the deadline to be the deadline for achieving total consistency, i.e. across receiving all reports, not just each one.

Reply:
The deadline is created at the start of the loop so is the same instant for all `rx.recv()` and will timeout the loop if we haven't successfully exited.